BLUESPAWN-client/src/user/CLI.cpp

```diff
@@ -7,82 +7,136 @@
-CLI::CLI() :
-	input{ input },
-	output{ output }
-{
-	//instances.emplace(std::pair(input, output), *this);
-}
-void SetConsoleColor(MessageColor color) {
+//Identifiers for different message types
+#define INFORM_ID L"[*]"
+#define ALERT_ID L"[!]"
+#define CONFIRM_ID L"[+]"
+#define SELECT_ID L"[?]"
+
+//Colors for different parts of messages
+const MessageColor ID_COLOR = MessageColor::BLUE;
+const MessageColor TEXT_COLOR = MessageColor::LIGHTGRAY;
+
+//Case-insensitive options for user confirmation.
+const std::set<std::wstring> affirmativeOptions = { L"yes", L"y" };
+const std::set<std::wstring> negativeOptions = { L"no", L"n" };
+const std::set<std::wstring> cancelOptions = { L"c", L"cancel" };
+
+const std::wstring descriptions[3] = {
+	L"[LOW]",
+	L"[MEDIUM]",
+	L"[HIGH]"
+};
+
+const MessageColor colors[3] = {
+	MessageColor::GREEN,
+	MessageColor::YELLOW,
+	MessageColor::RED
+};
+
+CLI::CLI() : hMutex{ CreateMutexW(nullptr, false, L"Local\\CLI-Mutex") } {}
+
+void Print(const std::wstring& wMessage, MessageColor color = TEXT_COLOR, bool newline=true){
 	HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
 	SetConsoleTextAttribute(hConsole, static_cast<WORD>(color));
+	if(newline){
+		std::wcout << wMessage << std::endl;
+	} else {
+		std::wcout << wMessage;
+	}
 }
 
-std::wstring CLI::GetUserSelection(const std::wstring& prompt, const std::vector<std::wstring>& options,
-	DWORD dwMaximumDelay) const {
-	SetConsoleColor(MessageColor::BLUE);
-	wcout << L"[?] ";
-	SetConsoleColor(MessageColor::WHITE);
+std::optional<std::wstring> GetInput(DWORD dwMaximumDelay){
+	std::wstring output{};
+	auto result = WaitForSingleObject(GetStdHandle(STD_INPUT_HANDLE), dwMaximumDelay);
+	if(result == WAIT_OBJECT_0){
+		std::wstring input{};
+		std::getline(std::wcin, input);
+		return input;
+	} else {
+		return std::nullopt;
+	}
+}
+
+const CLI CLI::instance{};
+
+const CLI& CLI::GetInstance(){
+	return instance;
+}
 
-	wstringstream stream;
-	stream << prompt << endl;
-	int i = 1;
+std::wstring CLI::GetUserSelection(const std::wstring& prompt, const std::vector<std::wstring>& options,
+	DWORD dwMaximumDelay, ImportanceLevel level) const {
+	auto mutex = AcquireMutex(hMutex);
+	Print(SELECT_ID, ID_COLOR, false);
+	Print(descriptions[static_cast<DWORD>(level)], colors[static_cast<DWORD>(level)], false);
+	Print(L" " + prompt);
 
+	int i = 0;
 	for (auto str : options) {
-		stream << i << L". " << str << endl;
-		i++;
+		i += 1;
+		Print(std::to_wstring(i), MessageColor::CYAN, false);
+		Print(L". " + str);
 	}
-	wcout << stream.str() << endl;
-	
-	int userIn;
-	while (true) {
-		wcin >> userIn;
-		if (userIn > 0 && userIn <= options.size()) {
-			return options[userIn-1];
-		}
-		else {
-			wcout << L"Please Enter a valid number between 1 and " << options.size() <<endl;
+	Print(L"Please enter a number 1 through " + std::to_wstring(i) + L" to continue. ", MessageColor::LIGHTGRAY, false);
+
+	while(true){
+		size_t userIn = 0;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The scope of the variable 'userIn' can be reduced.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5092780)